pkg/scheduler/algorithm/priorities/even_pods_spread.go

```diff
@@ -0,0 +1,209 @@
+	topologyPairToPodCounts map[topologyPair]*int64
+}
+
+func newTopologySpreadConstraintsMap() *topologySpreadConstraintsMap {
+	return &topologySpreadConstraintsMap{
+		nodeNameToPodCounts:     make(map[string]int64),
+		topologyPairToPodCounts: make(map[topologyPair]*int64),
+	}
+}
+
+// Note: the <nodes> passed in are the "filtered" nodes which have passed Predicates.
+// This function iterates <nodes> to filter out the nodes which don't have required topologyKey(s),
+// and initialize two maps:
+// 1) t.topologyPairToPodCounts: keyed with both eligible topology pair and node names.
+// 2) t.nodeNameToPodCounts: keyed with node name, and valued with a *int64 pointer for eligible node only.
+func (t *topologySpreadConstraintsMap) initialize(pod *v1.Pod, nodes []*v1.Node) {
+	constraints := getSoftTopologySpreadConstraints(pod)
+	for _, node := range nodes {
+		if !predicates.NodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
+			continue
+		}
+		for _, constraint := range constraints {
+			pair := topologyPair{key: constraint.TopologyKey, value: node.Labels[constraint.TopologyKey]}
+			if t.topologyPairToPodCounts[pair] == nil {
+				t.topologyPairToPodCounts[pair] = new(int64)
+			}
+		}
+		t.nodeNameToPodCounts[node.Name] = 0
+		// For those nodes which don't have all required topologyKeys present, it's intentional to keep
+		// those entries absent in nodeNameToPodCounts, so that we're able to score them to 0 afterwards.
+	}
+}
+
+// CalculateEvenPodsSpreadPriority computes a score by checking through the topologySpreadConstraints
+// that are with WhenUnsatisfiable=ScheduleAnyway (a.k.a soft constraint).
+// The function works as below:
+// 1) In all nodes, calculate the number of pods which match <pod>'s soft topology spread constraints.
+// 2) Group the number calculated in 1) by topologyPair, and sum up to corresponding candidate nodes.
+// 3) Finally normalize the number to 0~10. The node with the highest score is the most preferred.
+// Note: Symmetry is not applicable. We only weigh how incomingPod matches existingPod.
+// Whether existingPod matches incomingPod doesn't contribute to the final score.
+// This is different from the Affinity API.
+func CalculateEvenPodsSpreadPriority(pod *v1.Pod, nodeNameToInfo map[string]*schedulernodeinfo.NodeInfo, nodes []*v1.Node) (schedulerapi.HostPriorityList, error) {
+	result := make(schedulerapi.HostPriorityList, len(nodes))
+	// return if incoming pod doesn't have soft topology spread constraints.
+	constraints := getSoftTopologySpreadConstraints(pod)
+	if len(constraints) == 0 {
+		return result, nil
+	}
+
+	t := newTopologySpreadConstraintsMap()
+	t.initialize(pod, nodes)
+
+	allNodeNames := make([]string, 0, len(nodeNameToInfo))
+	for name := range nodeNameToInfo {
+		allNodeNames = append(allNodeNames, name)
+	}
+
+	errCh := schedutil.NewErrorChannel()
+	ctx, cancel := context.WithCancel(context.Background())
+	processAllNode := func(i int) {
```

Comment:
processNode seems to be better name.

Reply:
That's the old name. Here I want to highlight that this function works for **all** nodes, instead of each **candidate** node which has passed Predicates check.